### PR TITLE
Support mayactl volume/snapshot operation across namespaces

### PR DESCRIPTION
**Pull request by prateekpandey14**
1. This commit will enhance mayactl to perform volume create/delete/
   list/stats/info cmd for volumes other then default namespace.
2. Mayactl snapshot create/list/revert cmds across volume namespace.
fixes: https://github.com/openebs/openebs/issues/1510
fixes: https://github.com/openebs/openebs/issues/1600
**Volume stats  of percona namespaced volume:**
```sh
$ mayactl volume stats -n percona --volname pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db
Executing volume stats...
IQN     : iqn.2016-09.com.openebs.jiva:pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db
Volume  : pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db
Portal  : 10.111.86.151:3260
Size    : 5G

     Replica|    Status|   DataUpdateIndex|
            |          |                  |
   10.47.0.2|   Running|                 0|
   10.44.0.3|   Running|              2202|
   10.36.0.2|   Running|              2202|

----------- Performance Stats -----------
   r/s|   w/s|   r(MB/s)|   w(MB/s)|   rLat(ms)|   wLat(ms)|
     0|     0|     0.000|     0.000|      0.000|      0.000|

------------ Capacity Stats -------------
   Logical(GB)|   Used(GB)|
         0.352|      0.352|

```
**Volume Info of percona namespaced volume:**
```sh
mayactl volume info --namespace percona --volname pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db
volume namespace is percona

Portal Details :
---------------
IQN     :   iqn.2016-09.com.openebs.jiva:pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db
Volume  :   pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db
Portal  :   10.111.86.151:3260
Size    :   5G
Status  :   Running


Replica Details :
---------------- 
NAME                                                             ACCESSMODE      STATUS      IP            NODE 
-----                                                            -----------     -------     ---           ----- 
pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db-rep-87fc78659-8tzdr     NA              Running     10.47.0.2     kubeminion-02 
pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db-rep-87fc78659-cbfrn     RW              Running     10.44.0.3     kubeminion-01 
pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db-rep-87fc78659-dwxqb     RW              Running     10.36.0.2     kubeminion-03 

```
**Snapshot create of percona namespaced volume:**
```sh
$ mayactl snapshot --namespace percona create --volname pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db --snapname testsnap
Executing volume snapshot create...
Volume snapshot Successfully Created:pvc-186ee6ee-6987-11e8-ac26-02b983f0a4db

```
Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

### CODE REVIEW

File: cmd/mayactl/app/command/stats_helper_test.go [hunk 1]
```diff
@@ -81,12 +82,22 @@ func TestGetVolDetails(t *testing.T) {
 			err:  util.PageNotFound,
 			addr: "MAPI_ADDR",
 		},
+		"OpenebsNamespacedVolume": {
+			volumeName: "vol",
+			namespace:  "",
```
Comment (utkarshmani1997) [inline cmd/mayactl/app/command/stats_helper_test.go]: It will be good to verify one another test case for the any namespace other than  `""`.

Comment (prateekpandey14) [reply cmd/mayactl/app/command/stats_helper_test.go]: Its there.

File: cmd/mayactl/app/command/snapshot/snapshot.go [hunk 2]
```diff
@@ -32,6 +44,8 @@ func NewCmdSnapshot() *cobra.Command {
 		NewCmdSnapshotList(),
 		NewCmdSnapshotRevert(),
 	)
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace,
+		"namespace name, required if volume is in other then dafault namespace")
```
Comment (kmova) [inline cmd/mayactl/app/command/snapshot/snapshot.go]: _n: Fix this sentence please. Get this reviewed with @mahebbar

File: cmd/mayactl/app/command/snapshot/create.go [hunk 3]
```diff
@@ -44,17 +44,8 @@ import (
 }
 */
 
-// CmdSnaphotCreateOptions holds the options for snapshot
```
Comment (mahebbar) [inline cmd/mayactl/app/command/snapshot/create.go]: holds = contains

File: cmd/mayactl/app/command/snapshot/create.go [hunk 4]
```diff
@@ -65,34 +56,32 @@ func NewCmdSnapshotCreate() *cobra.Command {
...
 
 // RunSnapshotCreate does tasks related to mayaserver.
-func (c *CmdSnaphotCreateOptions) RunSnapshotCreate(cmd *cobra.Command) error {
+func (c *CmdSnaphotOptions) RunSnapshotCreate(cmd *cobra.Command) error {
 	fmt.Println("Executing volume snapshot create...")
 
-	resp := mapiserver.CreateSnapshot(c.volName, c.snapName)
+	resp := mapiserver.CreateSnapshot(c.volName, c.snapName, c.namespace)
 	if resp != nil {
 		return fmt.Errorf("Snapshot create failed: %v", resp)
```
Comment (mahebbar) [inline cmd/mayactl/app/command/snapshot/create.go]: create failed = creation failed

File: cmd/mayactl/app/command/snapshot/list.go [hunk 5]
```diff
@@ -39,24 +37,24 @@ func NewCmdSnapshotList() *cobra.Command {
...
-func (c *CmdSnaphotCreateOptions) ValidateList(cmd *cobra.Command) error {
-	if c.volName == "" {
+func (c *CmdSnaphotOptions) ValidateList(cmd *cobra.Command) error {
+	if len(c.volName) == 0 {
 		return errors.New("--volname is missing. Please specify an unique name")
 	}
 	return nil
 }
 
 // RunSnapshotList does tasks related to mayaserver.
```
Comment (mahebbar) [inline cmd/mayactl/app/command/snapshot/list.go]: does = performs

File: cmd/mayactl/app/command/snapshot/snapshot.go (hunk 2 above)
Comment (mahebbar) [inline cmd/mayactl/app/command/snapshot/snapshot.go]: required if a volume is not in the default namespace

File: cmd/mayactl/app/command/stats_helper.go [hunk 6]
```diff
@@ -42,11 +42,18 @@ func GetVolDetails(volName string, obj interface{}) error {
...
-	resp, err := client.Get(url)
 
+	req.Header.Set("namespace", namespace)
+
+	c := &http.Client{
+		Timeout: timeout,
+	}
+	resp, err := c.Do(req)
 	if err != nil {
 		fmt.Printf("Could not get response, found error: %v", err)
```
Comment (mahebbar) [inline cmd/mayactl/app/command/stats_helper.go]: Cannot get a response, error found:

File: cmd/mayactl/app/command/volume.go [hunk 7]
```diff
@@ -56,5 +68,10 @@ func NewCmdVolume() *cobra.Command {
 		NewCmdVolumeStats(),
 		NewCmdVolumeInfo(),
 	)
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace,
+		"namespace name, required if volume is in other then dafault namespace")
```
[the remark this answers is not in the document]
Comment (mahebbar) [reply cmd/mayactl/app/command/volume.go]: required if volume is not in the default namespace

File: cmd/mayactl/app/command/volume_create.go [hunk 8]
```diff
@@ -64,20 +57,21 @@ func NewCmdVolumeCreate() *cobra.Command {
 
 	cmd.Flags().StringVarP(&options.size, "size", "", options.size,
 		"volume capacity in GB (example: 10G) (default: 5G")
-
 	return cmd
 }
 
-// Run does tasks related to mayaserver.
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_create.go]: does tasks = performs tasks

File: cmd/mayactl/app/command/volume_create.go [hunk 9]
```diff
@@ -64,20 +57,21 @@ func NewCmdVolumeCreate() *cobra.Command {
...
-func (c *CmdVolumeCreateOptions) Validate(cmd *cobra.Command) error {
-	if c.volName == "" {
+// Validate varifies whether a volume name has been provided or not followed by
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_create.go]: varifies = verifies
has been provided = is provided

File: cmd/mayactl/app/command/volume_create.go [hunk 10]
```diff
@@ -64,20 +57,21 @@ func NewCmdVolumeCreate() *cobra.Command {
...
+// stats command, it returns nil and proceeds to execute the command if there is
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_create.go]: command. It returns...

File: cmd/mayactl/app/command/volume_create.go [hunk 11]
```diff
@@ -64,20 +57,21 @@ func NewCmdVolumeCreate() *cobra.Command {
...
+// no error and returns the error if it is missing.
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_create.go]: returns the error = returns an error

File: cmd/mayactl/app/command/volume_delete.go [hunk 12]
```diff
@@ -34,45 +33,29 @@ var (
...
 		"unique volume name.")
 	cmd.MarkPersistentFlagRequired("volname")
-
 	return cmd
 }
 
-//ValidateVolumeDelete validates the arguments passed
-func (c *CmdVolumeDeleteOptions) ValidateVolumeDelete(cmd *cobra.Command) error {
-	if c.volName == "" {
-		return errors.New("--volname is missing. Please specify an unique name")
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_delete.go]: an unique = a unique

File: cmd/mayactl/app/command/volume_delete.go [hunk 13]
```diff
@@ -34,45 +33,29 @@ var (
...
-	}
-	return nil
-}
-
 //RunVolumeDelete will initiate deletion of volume from maya-apiserver
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_delete.go]: will initiate the process of deleting a volume from...

Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_delete.go]: done.

File: cmd/mayactl/app/command/volume_delete.go (hunk 12 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_delete.go]: done.

File: cmd/mayactl/app/command/volume_create.go (hunk 11 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_create.go]: done.

File: cmd/mayactl/app/command/volume_create.go (hunk 10 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_create.go]: done.

File: cmd/mayactl/app/command/volume_create.go (hunk 9 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_create.go]: done.

File: cmd/mayactl/app/command/volume_info.go [hunk 14]
```diff
@@ -50,14 +49,8 @@ type ReplicaInfo struct {
 	NodeName   string
 }
 
-// CmdVolumeInfoOptions is used to store the value of flags used in the cli
-type CmdVolumeInfoOptions struct {
-	volName string
-}
-
 // NewCmdVolumeInfo shows info of OpenEBS Volume
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: shows = displays OpenEBS Volume information

File: cmd/mayactl/app/command/snapshot/snapshot.go (hunk 2 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/snapshot/snapshot.go]: done.

File: cmd/mayactl/app/command/volume_info.go [hunk 15]
```diff
@@ -50,14 +49,8 @@ type ReplicaInfo struct {
...
 func NewCmdVolumeInfo() *cobra.Command {
-	options := CmdVolumeInfoOptions{}
 	cmd := &cobra.Command{
 		Use:     "info",
 		Short:   "Displays the info of Volume",
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: Displays OpenEBS Volume information

File: cmd/mayactl/app/command/volume_info.go [hunk 16]
```diff
@@ -68,27 +61,18 @@ func NewCmdVolumeInfo() *cobra.Command {
...
 		},
 	}
-	cmd.Flags().StringVarP(&options.volName, "volname", "n", options.volName,
+	cmd.Flags().StringVarP(&options.volName, "volname", "", options.volName,
 		"unique volume name.")
-
 	return cmd
 }
 
-// Validate verifies the command whether volName is passed or not.
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: ...verifies if the volName command has passed.

File: cmd/mayactl/app/command/snapshot/list.go (hunk 5 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/snapshot/list.go]: done.

File: cmd/mayactl/app/command/snapshot/create.go (hunk 4 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/snapshot/create.go]: done.

File: cmd/mayactl/app/command/volume.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume.go]: done.

File: cmd/mayactl/app/command/snapshot/snapshot.go (hunk 2 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/snapshot/snapshot.go]: done.

File: cmd/mayactl/app/command/volume_info.go [hunk 17]
```diff
@@ -68,27 +61,18 @@ func NewCmdVolumeInfo() *cobra.Command {
...
-
 // TODO : Add more volume information
 // RunVolumeInfo runs info command and make call to DisplayVolumeInfo
-func (c *CmdVolumeInfoOptions) RunVolumeInfo(cmd *cobra.Command) error {
+func (c *CmdVolumeOptions) RunVolumeInfo(cmd *cobra.Command) error {
 	annotation := &Annotations{}
 	// GetVolumeAnnotation is called to get the volume controller's info such as
-	// controller's IP, status, iqn, replica IPs etc.
-	err := annotation.GetVolAnnotations(c.volName)
+	// controller's IP, status, iqn, replica IPs etc.\
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volume_info.go [hunk 18]
```diff
@@ -68,27 +61,18 @@ func NewCmdVolumeInfo() *cobra.Command {
...
+	cmd.Flags().StringVarP(&options.volName, "volname", "", options.volName,
 		"unique volume name.")
-
 	return cmd
 }
 
-// Validate verifies the command whether volName is passed or not.
-func (c *CmdVolumeInfoOptions) Validate(cmd *cobra.Command) error {
-	if c.volName == "" {
-		return errors.New("--volname is missing. Please try running [mayactl volume list] to see list of volumes.")
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: see the list of volumes.

File: cmd/mayactl/app/command/volume_info.go [hunk 19]
```diff
@@ -68,27 +61,18 @@ func NewCmdVolumeInfo() *cobra.Command {
...
-	}
-	return nil
-}
-
 // TODO : Add more volume information
 // RunVolumeInfo runs info command and make call to DisplayVolumeInfo
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: and makes a call...

File: cmd/mayactl/app/command/volume_info.go [hunk 20]
```diff
@@ -136,7 +120,7 @@ func updateReplicasInfo(replicaInfo map[int]*ReplicaInfo) error {
 
 // DisplayVolumeInfo displays the outputs in standard I/O.
 // Currently It displays volume access modes and target portal details only.
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: It = it

File: cmd/mayactl/app/command/volume_info.go [hunk 21]
```diff
@@ -136,7 +120,7 @@ func updateReplicasInfo(replicaInfo map[int]*ReplicaInfo) error {
...
-func (c *CmdVolumeInfoOptions) DisplayVolumeInfo(a *Annotations, collection client.ReplicaCollection) error {
+func (c *CmdVolumeOptions) DisplayVolumeInfo(a *Annotations, collection client.ReplicaCollection) error {
 	var (
 		// address, mode are used here as blackbox for the replica info
```
Comment (mahebbar) [inline cmd/mayactl/app/command/volume_info.go]: address and mode are...

File: cmd/mayactl/app/command/volume_info.go (hunk 14 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volume_info.go (hunk 15 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volume_info.go (hunk 20 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volume_info.go (hunk 21 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volume_stats.go [hunk 22]
```diff
@@ -156,7 +136,7 @@ func (c *CmdVolumeStatsOptions) RunVolumeStats(cmd *cobra.Command) error {
 // DisplayStats displays the volume stats as standard output and in json format.
 // By defaault it displays in standard output but if  flag json is passed it
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_stats.go]: done.

File: cmd/mayactl/app/command/volumes_list.go [hunk 23]
```diff
@@ -34,15 +34,8 @@ var (
 	`
 )
 
-// CmdVolumesListOptions captures the CLI flags
-type CmdVolumesListOptions struct {
-	volName string
-}
-
 // NewCmdVolumesList display status of OpenEBS Volume(s)
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volumes_list.go]: done.

File: cmd/mayactl/app/command/volumes_list.go [hunk 24]
```diff
@@ -34,15 +34,8 @@ var (
...
 func NewCmdVolumesList() *cobra.Command {
-	options := CmdVolumesListOptions{}
-
 	cmd := &cobra.Command{
 		Use:   "list",
 		Short: "Display status information about Volume(s)",
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volumes_list.go]: done..

File: pkg/client/mapiserver/snapshot.go [hunk 25]
```diff
@@ -132,7 +133,7 @@ func RevertSnapshot(volName string, snapName string) error {
 }
 
 // ListSnapshot list snapshots of volume by invoking the API call to m-apiserver
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/client/mapiserver/snapshot.go]: done.

File: cmd/mayactl/app/command/volume_info.go (hunk 18 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volume_info.go (hunk 19 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume_info.go]: done.

File: cmd/mayactl/app/command/volumes_list.go [hunk 26]
```diff
@@ -60,7 +53,7 @@ func NewCmdVolumesList() *cobra.Command {
 }
 
 //RunVolumesList will fetch the volumes from maya-apiserver
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volumes_list.go]: done.

File: pkg/client/mapiserver/snapshot.go [hunk 27]
```diff
@@ -85,7 +86,7 @@ func CreateSnapshot(volName string, snapName string) error {
 }
 
 // RevertSnapshot revert a snapshot of volume by invoking the API call to m-apiserver
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/client/mapiserver/snapshot.go]: done

**Pull request merged**